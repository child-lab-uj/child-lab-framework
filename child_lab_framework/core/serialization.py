from types import NoneType
from typing import Protocol, Self, TypeGuard

type Value = None | int | float | str | list[Value] | dict[str, Value]


class DeserializeError(Exception): ...


def assert_proper_serialization(data: dict[object, object]) -> None:
    for key, value in data.items():
        if not isinstance(key, str):
            raise DeserializeError(f'Non-string key found: {key}')

        if not is_allowed(value):
            raise DeserializeError(f'Illegal value found: {value}')


def is_allowed(value: object) -> TypeGuard[Value]:
    if type(value) in {NoneType, bool, int, float, str}:
        return True

    match value:
        case list(values):
            return all(map(is_allowed, values))

        case dict(entries):
            return is_serialization(entries)

        case _:
            return False


def is_serialization(data: dict[object, object]) -> TypeGuard[dict[str, Value]]:
    return all(isinstance(key, str) and is_allowed(value) for key, value in data.items())


class Serializable(Protocol):
    def serialize(self) -> dict[str, Value]: ...

    @classmethod
    def deserialize(cls, data: dict[str, Value]) -> Self: ...
