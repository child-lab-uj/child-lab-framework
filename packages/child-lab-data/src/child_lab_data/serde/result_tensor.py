import torch
from beartype import beartype
from jaxtyping import Float
from plum import dispatch, overload
from vpc import gaze, pose


@beartype
@overload
def serialize(item: pose.Result3d) -> Float[torch.Tensor, 'n_detections 74']:  # noqa: F811
    flat_boxes = item.boxes  # n_detections x 6
    flat_keypoints = item.keypoints.flatten(1, -1)  # n_detections x 17 * 4
    return torch.cat((flat_boxes, flat_keypoints), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[pose.Result3d],
    item: Float[torch.Tensor, 'n_detections 74'],
) -> pose.Result3d:
    boxes = item[:, :6]
    keypoints = item[:, 6:].unflatten(-1, (17, 4))
    return pose.Result3d(boxes, keypoints, item.device)


@beartype
@overload
def serialize(item: pose.Result) -> Float[torch.Tensor, 'n_detections 56']:  # noqa: F811
    flat_boxes = item.boxes  # n_detections x 5
    flat_keypoints = item.keypoints.flatten(1, -1)  # n_detections x 17 * 3
    return torch.cat((flat_boxes, flat_keypoints), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[pose.Result],
    item: Float[torch.Tensor, 'n_detections 56'],
) -> pose.Result:
    boxes = item[:, :6, :]
    keypoints = item[:, 6:, :].unflatten(-1, (17, 3))
    return pose.Result(boxes, keypoints, item.device)


@beartype
@overload
def serialize(item: gaze.Result3d) -> Float[torch.Tensor, 'n_detections 4 3']:  # noqa: F811
    return torch.cat((item.eyes, item.directions), dim=1)


@beartype
@overload
def deserialize(  # noqa: F811
    ty: type[gaze.Result3d],
    item: Float[torch.Tensor, 'n_detections 4 3'],
) -> gaze.Result3d:
    eyes = item[:, :2, :]
    directions = item[:, 2:, :]
    return gaze.Result3d(eyes, directions, item.device)


@dispatch
def serialize(item: object) -> torch.Tensor: ...  # noqa: F811


@dispatch
def deserialize(ty: type, item: object) -> object: ...  # noqa: F811
