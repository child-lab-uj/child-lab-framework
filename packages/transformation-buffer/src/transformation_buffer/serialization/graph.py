from typing import Any

import serde
from plum import dispatch
from rustworkx import PyDiGraph

from transformation_buffer.transformation import Transformation


def init() -> None:
    serde.add_serializer(Serializer())
    serde.add_deserializer(Deserializer())


class Serializer:
    @dispatch
    def serialize(self, value: PyDiGraph) -> dict[str, Any]:
        return serialize_graph(value)


class Deserializer:
    @dispatch
    def deserialize(self, cls: type[PyDiGraph], value: Any) -> PyDiGraph:
        return deserialize_graph(value)


def serialize_graph(graph: PyDiGraph) -> dict[str, Any]:
    nodes = [(index, graph.get_node_data(index)) for index in graph.node_indices()]
    edges = [
        (from_to, serde.to_dict(graph.get_edge_data(*from_to)))
        for from_to in graph.edge_list()
    ]

    return {
        'check_cycle': graph.check_cycle,
        'multigraph': graph.multigraph,
        'nodes': nodes,
        'edges': edges,
    }


# INVARIANT: Properly deserializes only graphs with Transformations as edge poayloads.
def deserialize_graph(value: Any) -> PyDiGraph:
    match value:
        case {
            'check_cycle': bool(check_cycle),
            'multigraph': bool(multigraph),
            'nodes': list(nodes),
            'edges': list(edges),
        }:
            graph = PyDiGraph(
                check_cycle,
                multigraph,
                node_count_hint=len(nodes),
                edge_count_hint=len(edges),
            )

            old_node_indices_to_new = {}

            for old_index, node in sorted(nodes, key=lambda entry: entry[0]):
                new_index = graph.add_node(node)
                old_node_indices_to_new[old_index] = new_index

            for (old_a, old_b), edge in edges:
                new_a = old_node_indices_to_new[old_a]
                new_b = old_node_indices_to_new[old_b]

                # TODO: Unhack
                graph.add_edge(new_a, new_b, serde.from_dict(Transformation, edge))

            return graph

        case _:
            raise serde.SerdeError('')
