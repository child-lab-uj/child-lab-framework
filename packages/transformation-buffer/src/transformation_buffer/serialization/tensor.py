from typing import Any

import serde
import torch
from plum import dispatch


def init() -> None:
    serde.add_serializer(Serializer())
    serde.add_deserializer(Deserializer())


class Serializer:
    @dispatch
    def serialize(self, value: torch.Tensor) -> dict[str, Any]:
        return serialize_tensor(value)


class Deserializer:
    @dispatch
    def deserialize(self, cls: type[torch.Tensor], value: Any) -> torch.Tensor:
        return deserialize_tensor(value)


def serialize_tensor(tensor: torch.Tensor) -> dict[str, Any]:
    return {'data': tensor.tolist(), 'dtype': str(tensor.dtype)}


def deserialize_tensor(data: Any) -> torch.Tensor:
    match data:
        case {'data': list(data), 'dtype': str(serialized_dtype)}:
            dtype: torch.dtype = eval(serialized_dtype, None, {'torch': torch})
            return torch.tensor(data, dtype=dtype)

        case _:
            raise serde.SerdeError('')
