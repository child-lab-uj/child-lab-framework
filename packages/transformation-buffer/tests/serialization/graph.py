import pytest
import serde
from rustworkx import PyDiGraph
from serde.json import from_json, to_json
from serde.yaml import to_yaml
from syrupy.assertion import SnapshotAssertion
from transformation_buffer.serialization.graph import deserialize_graph, serialize_graph


# A wrapper class to test field serialization.
# `PyDiGraph` cannot be serialized to any concrete format itself.
@serde.serde
class WrapperClass[N, E]:
    graph: PyDiGraph[N, E] = serde.field(
        serializer=serialize_graph,
        deserializer=deserialize_graph,
    )


@pytest.fixture
def empty_graph() -> PyDiGraph[int, int]:
    return PyDiGraph[int, int]()


@pytest.fixture
def simple_graph() -> PyDiGraph[str, str]:
    graph = PyDiGraph[str, str]()

    a = graph.add_node('a')
    b = graph.add_node('b')
    c = graph.add_node('c')
    d = graph.add_node('d')
    e = graph.add_node('e')

    graph.add_edge(a, b, 'edge1')
    graph.add_edge(a, c, 'edge2')
    graph.add_edge(d, e, 'edge3')

    return graph


def test_serialize_with_empty_graph_json(
    empty_graph: PyDiGraph[int, int],
    snapshot: SnapshotAssertion,
) -> None:
    obj = WrapperClass(empty_graph)
    serialized_obj = to_json(obj)
    assert serialized_obj == snapshot


def test_serialize_with_simple_graph_json(
    simple_graph: PyDiGraph[str, str],
    snapshot: SnapshotAssertion,
) -> None:
    obj = WrapperClass(simple_graph)
    serialized_obj = to_json(obj)
    assert serialized_obj == snapshot


def test_deserialize_with_empty_graph_json() -> None:
    data = """
        {
            "graph": {
                "check_cycle": false,
                "multigraph": true,
                "nodes": [],
                "edges": []
            }
        }
    """

    graph = from_json(WrapperClass[int, int], data).graph

    assert not graph.check_cycle
    assert graph.multigraph
    assert graph.nodes() == []
    assert graph.edges() == []


def test_deserialize_with_simple_graph_json() -> None:
    data = """
        {
            "graph": {
                "check_cycle": false,
                "multigraph": true,
                "nodes": [
                    [0, "a"],
                    [1, "b"],
                    [2, "c"],
                    [3, "d"],
                    [4, "e"]
                ],
                "edges": [
                    [[0, 1], "edge1"],
                    [[0, 2], "edge2"],
                    [[3, 4], "edge3"]
                ]
            }
        }
    """

    graph = from_json(WrapperClass[str, str], data).graph

    assert not graph.check_cycle
    assert graph.multigraph
    assert graph.nodes() == ['a', 'b', 'c', 'd', 'e']
    assert graph.edges() == ['edge1', 'edge2', 'edge3']


def test_serialize_with_empty_graph_yaml(
    empty_graph: PyDiGraph[int, int],
    snapshot: SnapshotAssertion,
) -> None:
    obj = WrapperClass(empty_graph)
    serialized_obj = to_yaml(obj)
    assert serialized_obj == snapshot
