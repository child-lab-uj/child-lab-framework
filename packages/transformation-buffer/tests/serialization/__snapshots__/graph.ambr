# serializer version: 1
# name: test_serialize_with_empty_graph_json
  '{"graph":{"check_cycle":false,"multigraph":true,"nodes":[],"edges":[]}}'
# ---
# name: test_serialize_with_empty_graph_yaml
  '''
  graph:
    check_cycle: false
    edges: []
    multigraph: true
    nodes: []
  
  '''
# ---
# name: test_serialize_with_simple_graph_json
  '{"graph":{"check_cycle":false,"multigraph":true,"nodes":[[0,"a"],[1,"b"],[2,"c"],[3,"d"],[4,"e"]],"edges":[[[0,1],"edge1"],[[0,2],"edge2"],[[3,4],"edge3"]]}}'
# ---
